easybuild/easyconfigs/d/dpcpp/dpcpp-6.0.0-GCCcore-13.3.0.eb

```
@@ -0,0 +1,99 @@
+# version of CUDA CUPTI library which would reduce portability, as well as
+# fixing an issue with finding libcuda.so
+patches = ['dpcpp-6.0.0-cmake-cuda-deps.patch']
+
+checksums = [
+    {'v6.0.0.tar.gz': 'f7a30b3075e2690b207063bb66504ebcb466ddf10604f6467b5a1785eb022ce8'},
+    {'dpcpp-6.0.0-cmake-cuda-deps.patch': '37cb78ac15d351dc233a19856454b3bc235ecec6a6d71d59d40805bc8ff7b279'},
+]
+
+builddependencies = [
+    ('CMake', '3.29.3'),
+    ('Ninja', '1.12.1'),
+    ('Python', '3.12.3'),
+    # CUDA >= 11.7 is recommended for the NVIDIA GPU backend in DPC++.
+    # Since CUDA is backwards compatible, any higher version will work
+    # at runtime with this build. DPC++ can also be used without CUDA
+    # to compile for other backends, hence no runtime dependency on CUDA.
+    ('CUDA', '11.7.0', '', SYSTEM),
+]
+dependencies = [
+    ('binutils', '2.42'),
+    ('hwloc', '2.10.0'),
+]
+
+configure_without_installdir = True
+configure_cmd = ' '.join([
+    # Use the configure script provided in the source
+    'python %(start_dir)s/buildbot/configure.py',
+    '-o %(builddir)s/build -t Release --cmake-gen Ninja',
+    '--cmake-opt="-DCMAKE_INSTALL_PREFIX=%(installdir)s"',
+    # Include OpenMP libraries in the build
+    '--llvm-external-projects "openmp"',
+    # Enable the SYCL CUDA backend
+    '--cuda --cmake-opt="-DCUDA_TOOLKIT_ROOT_DIR=$EBROOTCUDA"',
+    # Disable tracing which would otherwise introduce a dependency on a specific
+    # version of CUDA (CUPTI) and reduce portability
+    '--cmake-opt="-DSYCL_ENABLE_XPTI_TRACING=OFF"',
+])
+
+build_cmd_targets = 'deploy-sycl-toolchain omp'
+build_dir = '%(builddir)s/build'
+buildopts = '-C ' + build_dir + ' ' + build_cmd_targets
+
+install_target_subdir = '%(builddir)s/build/install'
+installopts = '-C ' + build_dir
+
+# Ensure DPC++ picks up libstdc++ from the GCCcore module and not from system
+postinstallcmds = [
+    'echo "--gcc-toolchain=$EBROOTGCCCORE" > %(installdir)s/bin/clang++.cfg',
+    'echo "--gcc-toolchain=$EBROOTGCCCORE" > %(installdir)s/bin/clang.cfg'
+]
+
+# OpenMP libs install in lib/<target-triple>
+modextrapaths = {
+    'LIBRARY_PATH': 'lib/%(arch)s-unknown-linux-gnu',
+    'LD_LIBRARY_PATH': 'lib/%(arch)s-unknown-linux-gnu',
+}
+
+sanity_check_paths = {
+    'files': ['bin/clang', 'bin/clang++', 'bin/sycl-ls',
+              'include/sycl/sycl.hpp', 'lib/libsycl.so', 'lib/libur_loader.so',
+              'lib/libur_adapter_cuda.so', 'lib/libur_adapter_level_zero.so',
+              'lib/libur_adapter_opencl.so', 'lib/libze_loader.so',
+              'lib/libOpenCL.so', 'lib/%(arch)s-unknown-linux-gnu/libomp.so'],
+    'dirs': ['bin', 'include', 'lib']
+}
+sanity_check_commands = [
+    'clang --version', 'clang++ --version', 'sycl-ls',
+    'echo "#include <sycl/sycl.hpp>" | clang++ -x c++ -fsycl ' +
+    '-fsycl-targets=spir64,spir64_x86_64,intel_gpu_pvc,nvidia_gpu_sm_60 ' +
+    '-nocudalib -o /dev/null -c -'
+]
```

```suggestion
]

moduleclass = 'compiler'
```
Note that this choice of `moduleclass` does have an impact in EasyBuild, particularly for the case of a hierarchical module naming scheme. That only really matters once you try to use `dpcpp` as a toolchain in an EasyBuild sense as it is incomplete (it does not offer a Fortran compiler)...but I wouldn't hold back the PR for this issue.  Maybe @boegel has a suggestion here?

---

Is it a requirement for a `compiler` class in EasyBuild to provide a Fortran compiler? Indeed DPC++ does not provide it.

It is a fully-fledged C/C++/CUDA compiler using the upstream Clang components, plus it is a SYCL compiler on top. I certainly foresee `dpcpp` being used as the compiler to build other EasyBuild modules. For example, one could consider building [GROMACS](https://manual.gromacs.org/2025.1/install-guide/index.html#sycl-gpu-acceleration-for-intel-gpus) with SYCL GPU acceleration. I would personally also like to see projects like [oneMath](https://github.com/uxlfoundation/oneMath) coming to EasyBuild in the future, built with `dpcpp`.

---

Right now, a fully fledged toolchain does require a Fortran compiler, `gfortran` could work because you build on GCCcore and that is what we have done in the past for older Clang versions (we've been waiting a long time for `flang` to mature). I wonder if `ifx` is an option? Either way, you would need to define one as part of your toolchain definition.

Looks like about to see quite a few Clang-based toolchains be defined (`aocc`, `dpcpp`, `ROCm` and upstream `clang`). @boegel we will need a strategy here otherwise it will lead to an explosion of (largely duplicate) easyconfig PRs...

---

I'm afraid `ifx` is not an option as we want to keep the `dpcpp` toolchain FOSS and cross-platform, whereas `ifx` is proprietary and only for x86 hosts. I would be happy setting `gfortran` as the Fortran compiler here (if you could advise how to do this), or I could even try compiling the in-tree LLVM Flang as part of this easyconfig. However, the latter is not tested or maintained in the DPC++ repository - we only pull down from upstream llvm-project as-is. It might still work just fine.

In any case, I'll wait for feedback from @boegel before taking any actions.

---

There's nothing you need to do as part of this PR, toolchain definition happens in the EasyBuild framework repository (for example see https://github.com/easybuilders/easybuild-framework/blob/develop/easybuild/toolchains/foss.py which is top level of the GCC-based FOSS toolchain we currently have)